pkg/license/aggregator.go

```diff
@@ -0,0 +1,216 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package license
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	apmv1 "github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/apmserver"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
+	essettings "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/settings"
+	kbconfig "github.com/elastic/cloud-on-k8s/pkg/controller/kibana/config"
+	kbpod "github.com/elastic/cloud-on-k8s/pkg/controller/kibana/pod"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// Aggregator aggregates the total of resources of all Elastic managed components
+type Aggregator struct {
+	client k8s.Client
+}
+
+type aggregate func() (resource.Quantity, error)
+
+// AggregateMemory aggregates the total memory of all Elastic managed components
+func (a Aggregator) AggregateMemory() (resource.Quantity, error) {
+	var totalMemory resource.Quantity
+
+	for _, f := range []aggregate{
+		a.aggregateElasticsearchMemory,
+		a.aggregateKibanaMemory,
+		a.aggregateApmServerMemory,
+	} {
+		memory, err := f()
+		if err != nil {
+			return resource.Quantity{}, err
+		}
+		totalMemory.Add(memory)
+	}
+
+	return totalMemory, nil
+}
+
+func (a Aggregator) aggregateElasticsearchMemory() (resource.Quantity, error) {
+	var esList esv1.ElasticsearchList
+	err := a.client.List(&esList)
+	if err != nil {
+		return resource.Quantity{}, err
```

Comment:
nit: might be worth using errors.Wrap() here (and in the other funcs) so we can have a stack trace and know where it's coming from